File: src/net/merayen/elastic/backend/architectures/local/nodes/sample_1/LProcessor.java
package net.merayen.elastic.backend.architectures.local.nodes.sample_1;

import net.merayen.elastic.backend.architectures.local.LocalProcessor;
import net.merayen.elastic.backend.architectures.local.lets.*;
import net.merayen.elastic.util.Postmaster;

public class LProcessor extends LocalProcessor {
	@Override
	protected void onInit() {}

	@Override
	protected void onPrepare() {}

	@Override
	protected void onProcess() {
		Inlet control = getInlet("control");
		AudioOutlet out = (AudioOutlet)getOutlet("out");

		if(out != null) {
			if(control instanceof MidiInlet) {
				// Playback is controlled by midi
				processWithMidi((MidiInlet)control, out);
			} else if(control instanceof SignalInlet) {
				// Playback is controlled by a signal (mostly just the speed)
				// Mostly appropriate when run inside a poly-node, that controls when it plays back
				processWithSignal((SignalInlet)control, out);
			} else {
				control.read = control.outlet.written;
			}
		} else if(control != null){
			control.read = control.outlet.written;
		}
	}

	private void processWithMidi(MidiInlet control, AudioOutlet out) {
		int start = out.written;
		int stop = control.outlet.written;

		control.read = stop;
		out.written = stop;
	}

	private void processWithSignal(SignalInlet control, AudioOutlet out) {
		int start = out.written;
		int stop = control.outlet.written;

		control.read = stop;
		out.written = stop;
	}

	@Override
	protected void onMessage(Postmaster.Message message) {}

	@Override
	protected void onDestroy() {}
}

File: src/main/kotlin/net/merayen/elastic/backend/architectures/local/nodes/value_1/LProcessor.kt
package net.merayen.elastic.backend.architectures.local.nodes.value_1

import net.merayen.elastic.backend.architectures.local.LocalProcessor
import net.merayen.elastic.backend.architectures.local.lets.SignalOutlet

class LProcessor : LocalProcessor() {
	private var currentTime = 0.0
	private var done = false

	override fun onProcess() {
		val elapsed = getOutlet("elapsed") as? SignalOutlet
		val beat = getOutlet("beat") as? SignalOutlet

		if (elapsed != null) { // Time in seconds sine the
			var t = currentTime
			val increase = 1.0 / sampleRate
			for (i in elapsed.signal.indices) {
				elapsed.signal[i] = t.toFloat()
				t += increase
			}

			currentTime = t

			elapsed.push()
		}

		if (beat != null) {
			val parentGroup = localNode.parentGroupNode
			var beatPosition = parentGroup.getBeatPosition()
			val bpm = parentGroup.getCurrentFrameBPM()
			val increment = (bpm / 60.0) / sampleRate

			for (i in beat.signal.indices) {
				beat.signal[i] = beatPosition.toFloat()
				beatPosition += increment
			}

			beat.push()
		}

		done = true
	}

	override fun onPrepare() {
		done = false
	}

	override fun onInit() {}
	override fun onDestroy() {}
}